src/jltypes.c

```diff
@@ -556,6 +556,33 @@ static void isort_union(jl_value_t **a, size_t len) JL_NOTSAFEPOINT
     }
 }
 
+static int simple_subtype(jl_value_t *a, jl_value_t *b, int hasfree)
+{
+    if (a == jl_bottom_type || b == (jl_value_t*)jl_any_type)
+        return 1;
+    if (jl_egal(a, b))
+        return 1;
+    if (hasfree == 0) {
+        // issue #24521: don't merge Type{T} where typeof(T) varies
+        return !(jl_is_type_type(a) && jl_is_type_type(b) &&
+             jl_typeof(jl_tparam0(a)) != jl_typeof(jl_tparam0(b))) &&
+             jl_subtype(a, b);
+    }
+    if (jl_is_typevar(a)) {
+        jl_value_t *na = ((jl_tvar_t*)a)->ub;
+        hasfree &= jl_has_free_typevars(na);
+        return simple_subtype(na, b, hasfree);
```

Comment:
In `Tuple{Union{Integer,T}, T} where T`, would this cause the result to become `Tuple{T, T} where T`, instead of the correct result with a switched Union:

```
julia> (Union{Tuple{Integer,T}, Tuple{T, T}} where T) == (Tuple{Union{Integer,T}, T} where T)
true
```

Reply:
Well it turns out this branch might affect method definition.
On master
```julia
julia> f(x::Union{AbstractArray,T}) where {T<:Vector} = T
f (generic function with 1 method)

julia> f([1])
ERROR: UndefVarError: `T` not defined
Stacktrace:
 [1] f(x::Vector{Int64})
   @ Main .\REPL[3]:1
 [2] top-level scope
   @ REPL[4]:1
```
This PR
```julia
julia> f(x::Union{AbstractArray,T}) where {T<:Vector} = T
WARNING: method definition for f at REPL[1]:1 declares type variable T but does not use it.
f (generic function with 1 method)

julia> f([1])
ERROR: UndefVarError: `T` not defined in static parameter matching
Suggestion: run Test.detect_unbound_args to detect method arguments that do not fully constrain a type parameter.
Stacktrace:
 [1] f(x::Vector{Int64})
   @ Main .\REPL[1]:1
 [2] top-level scope
   @ REPL[2]:1
```
Not sure if this is a regression.

Reply:
> Not sure if this is a regression.

IMO this is an improvement, not a regression, because it's better to throw than be ambiguous.

Reply:
I.e., you may have fixed a correctness bug?

Reply:
@N5N3 That is an interesting observation. It is sort of awkward that whether we catch that issue at definition or runtime depends somewhat on what normalizations are implemented for Union. I think I am okay with that change however. It seems fairly unlikely packages would have written that, and the solution is quite simple (just don't write that TypeVar in that place to fix the warning). In particular, it was already doing a similar normalization in several cases, so it is at least not an unexpected new behavior:

```
(on master)

julia> Union{Integer,T} where T<:Integer
Union{Integer, T} where T<:Integer

julia> Union{Any,T} where T<:Integer
Any # not a Union or TypeVar

julia> Union{Int64, T} where T<:Integer
Union{Int64, T} where T<:Integer

julia> Union{Union{},T} where T<:Integer
Integery # not a Union or TypeVar
```

No, this was not fixing a correctness bug. There is no ambiguity there.

Reply:
Unrelatedly, this PR is now very close to the unrelated issue encountered in https://github.com/JuliaLang/julia/issues/51637#issuecomment-1879245236, where it may error when the normalization fails, instead of simply not doing this normalization, if you are interested in taking a look at how it can avoid throwing the exception from normalization in the example there (for a later PR).